### PR TITLE
Fix handling of essential DOFs in GeometricMultigrid

### LINKED ISSUE #4002: Geometric Multigrid Convergence issues when operator is scaled by a constant
Posted about this in the slack channel, but when I run ./ex26 with all default parameters, I achieve convergence in 5 iterations. However, when I change the coefficient for the diffusion from 1 to 1000, I get that it takes 55 iterations. 

Doesn't this simply scale the matrix? The condition number should be the same, right? Why would it take so much longer?

Thank you,
Justin

### PR DESCRIPTION
**Pull request by pazner**
This fixes a bug in the multigrid solver by representing the prolongations by `RectangularConstrainedOperator`. If the prolongation operators are not constrained, then their transpose will have nontrivial action on the essential DOFs, resulting in degraded convergence.

Resolves #4002.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4003](https://github.com/mfem/mfem/pull/4003) | @pazner | @v-dobrev | @tarikdzanic + @sebastiangrimberg | 12/1/23 | 2/11/24 | 2/13/24 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: fem/multigrid.hpp [hunk 1]
```diff
@@ -140,7 +140,7 @@ class Multigrid : public MultigridBase
 
 public:
    /// Constructs an empty multigrid hierarchy
-   Multigrid();
+   Multigrid() { }
```
Comment (sebastiangrimberg) [inline fem/multigrid.hpp:143]: Maybe just

```suggestion
   Multigrid() = default;
```

?

**Review by sebastiangrimberg — APPROVED** [review]
Yep, I agree this is the correct fix and it resolves the issue https://github.com/mfem/mfem/issues/4002 in my tests. Just one minor comment on a default constructor.

**Comment by jtristano** [comment]
@pazner , @v-dobrev  - 
Thank you for the quick fix! 
When we run this with our elasticity parameters, we get the following
E=1, E>>1 -- good convergence ~10-18 iterations
E << 1, -- poor convergence rates 50-300 iterations

So I'm wondering why things head south when E << 1?  

We are also trying cases where Emax/Emin > 1e3 and we see similar behavior to when E << 1

Thanks
Joe

**Comment by pazner** [comment]
@jtristano, since you are solving elasticity, could it be that for $E \ll 1$ your material is nearly incompressible? In that case, I would expect worse convergence.

I also think that if the ratio $E_{\max} / E_{\min}$ is very large, then you will probably needed a stronger smoother to get robust convergence.

**Comment by tzanio** [comment]
ping: @tarikdzanic

**Review by tarikdzanic — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by v-dobrev** [comment]
Re-merged in `next` for testing...

**Review by joconnor819 — APPROVED** [review]
Thank y'all so much for these changes - this has helped us very much

**Comment by v-dobrev** [comment]
Re-merged in `next` for testing...

**Review by sebastiangrimberg — APPROVED** [review]

**Review by tarikdzanic — APPROVED** [review]

**Pull request merged**